src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -0,0 +1,272 @@
+            }
+            if (diagnoticsEnabled)
+            {
+                // Non-inline
+                RecordBeginRequestDiagnostics(httpContext, startTimestamp);
+            }
+
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void RequestEnd(HttpContext httpContext, Exception exception, HostingApplication.Context context)
+        {
+            // Local cache items resolved multiple items, in order of use so they are primed in cpu pipeline when used
+            var startTimestamp = context.StartTimestamp;
+
+            // If startTimestamp is 0, don't call GetTimestamp, likely don't need the value
+            var currentTimestamp = (startTimestamp != 0) ? Stopwatch.GetTimestamp() : 0;
+
+            // To keep the hot path short we defer logging to non-inlines
+            if (exception == null)
+            {
+                // No exception was thrown, request was sucessful
+                if (_diagnosticSource.IsEnabled(DiagnosticsEndRequestKey))
+                {
+                    // Diagnostics is enabled for EndRequest, but it may not be for BeginRequest
+                    // so call GetTimestamp if currentTimestamp is zero (from above)
+                    RecordEndRequestDiagnostics(
+                        httpContext,
+                        (currentTimestamp != 0) ? currentTimestamp : Stopwatch.GetTimestamp());
+                }
+            }
+            else
+            {
+                // Exception was thrown from request
+                if (_diagnosticSource.IsEnabled(DiagnosticsUnhandledExceptionKey))
+                {
+                    // Diagnostics is enabled for UnhandledException, but it may not be for BeginRequest
+                    // so call GetTimestamp if currentTimestamp is zero (from above)
+                    RecordUnhandledExceptionDiagnostics(
+                        httpContext,
+                        (currentTimestamp != 0) ? currentTimestamp : Stopwatch.GetTimestamp(),
+                        exception);
+                }
+
+                if (context.EventLogEnabled)
+                {
+                    // Non-inline
+                    HostingEventSource.Log.UnhandledException();
+                }
+            }
+
+            var activity = context.Activity;
+            // Always stop activity if it was started
+            if (activity != null)
+            {
+                _diagnosticSource.StopActivity(activity, new { HttpContext = httpContext });
+            }
+
+            // If startTimestamp was 0, then Information logging wasn't enabled at for this request (and calcuated time will be wildly wrong)
+            // Is used as proxy to reduce calls to virtual: _logger.IsEnabled(LogLevel.Information)
+            if (startTimestamp != 0)
+            {
+                // Non-inline
+                LogRequestFinished(httpContext, startTimestamp, currentTimestamp);
```

Comment:
could you log it before activity is stopped?
It could be quite useful for custom ILogger implementation to use Activity.Current to augment logs